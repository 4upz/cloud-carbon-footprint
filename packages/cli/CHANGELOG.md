# cloud-carbon-footprint

## 0.0.1

### Patch Changes

- Updated dependencies [95f326a]
  - @cloud-carbon-footprint/core@0.0.1

## 0.0.0

### Major Changes

- b6464f4: initialize changesets
