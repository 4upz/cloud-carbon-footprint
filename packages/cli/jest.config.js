/*
 * Â© 2021 Thoughtworks, Inc.
 */

// eslint-disable-next-line @typescript-eslint/no-var-requires
const baseConfig = require('../../jest.base.config')

module.exports = {
  ...baseConfig,
  coverageThreshold: {
    global: {
      statements: 97,
      branches: 90,
      functions: 91,
      lines: 97,
    },
  },
  coveragePathIgnorePatterns: ['<rootDir>/src/CliPrompts.ts'],
  testPathIgnorePatterns: ['<rootDir>/src/__tests__/fixtures'],
  modulePathIgnorePatterns: ['index.ts'],
}
