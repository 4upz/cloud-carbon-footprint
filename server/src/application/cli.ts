import { program } from 'commander'
import { App } from './App'
import * as process from 'process'
import { RawRequest } from './EstimationRequest'
import EmissionsTable from './EmissionsTable'
import CliPrompts from './CliPrompts'

export default async function cli(argv: string[] = process.argv) {
  program
    .option('-s, --startDate <string>', 'Start date in ISO format')
    .option('-e, --endDate <string>', 'End date in ISO format')
    .option('-r, --region <string>', 'AWS region to analyze')
    .option('-i, --interactive', 'Use interactive CLI prompts')

  program.parse(argv)

  let startDate, endDate
  let region

  if (program.interactive) {
    ;[startDate, endDate, region] = await CliPrompts()
  } else {
    startDate = program.startDate
    endDate = program.endDate
    region = program.region
  }
  const estimationRequest: RawRequest = { startDate, endDate, region }
  return await new App().getEstimate(estimationRequest).then(EmissionsTable)
}
