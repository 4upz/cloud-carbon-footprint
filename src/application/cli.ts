import { program } from 'commander'
import { App } from './App'
import * as process from 'process'
import { RawRequest } from './EstimationRequest'
import EmissionsTable from './EmissionsTable'
import CliPrompts from './CliPrompts'

export default async function cli(argv: string[] = process.argv) {
  program
    .option('-s, --startDate <string>', 'Start date in ISO format')
    .option('-e, --endDate <string>', 'End date in ISO format')
    .option('-i, --interactive', 'Use interactive CLI prompts')

  program.parse(argv)

  let startDate, endDate

  if (program.interactive) {
    ;[startDate, endDate] = await CliPrompts()
  } else {
    startDate = program.startDate
    endDate = program.endDate
  }
  const estimationRequest: RawRequest = { startDate, endDate }
  return await new App().getEstimate(estimationRequest).then(EmissionsTable)
}
