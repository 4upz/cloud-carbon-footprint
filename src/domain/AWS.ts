import CloudService from './CloudService'
import EBS from '../services/EBS'
import S3 from '../services/S3'

export default function AWS(): CloudService[] {
  return [new EBS(), new S3()]
}
