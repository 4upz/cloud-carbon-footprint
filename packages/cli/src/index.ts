/*
 * Â© 2021 Thoughtworks, Inc. All rights reserved.
 */

import cli from './cli'
import * as console from 'console'

cli().then(console.log)
