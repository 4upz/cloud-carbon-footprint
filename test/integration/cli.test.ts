import cli from '../../src/application/cli'
import AWSMock from 'aws-sdk-mock'
import AWS from 'aws-sdk'

beforeAll(() => {
  AWSMock.setSDKInstance(AWS)

  AWSMock.mock(
    'CostExplorer',
    'getCostAndUsage',
    (params: AWS.CostExplorer.GetCostAndUsageRequest, callback: (a: Error, response: any) => any) => {
      callback(null, {
        ResultsByTime: [
          {
            Estimated: false,
            Groups: [],
            TimePeriod: {
              End: '2020-06-28',
              Start: '2020-06-27',
            },
            Total: {
              UsageQuantity: {
                Amount: '1.0',
                Unit: 'GB-Month',
              },
            },
          },
          {
            Estimated: false,
            Groups: [],
            TimePeriod: {
              End: '2020-06-29',
              Start: '2020-06-28',
            },
            Total: {
              UsageQuantity: {
                Amount: '2.0',
                Unit: 'GB-Month',
              },
            },
          },
        ],
      })
    },
  )
})

afterAll(() => {
  AWSMock.restore()
})

describe('cli', () => {
  test('cli integration test', async () => {
    const result = await cli(['executable', 'file', '--startDate', '2020-07-10', '--endDate', '2020-07-13'])

    expect(result).toEqual(
      '| Date (UTC)        | EBS Wattage       | EBS CO2e Emissions          \n' +
        '| 2020-06-27        | 0.86 Watts        | 0.000611 Kg CO2e            \n' +
        '| 2020-06-28        | 1.73 Watts        | 0.001222 Kg CO2e            \n' +
        '| Total             | 2.59 Watts        | 0.001833 Kg CO2e            ',
    )
  })
})
