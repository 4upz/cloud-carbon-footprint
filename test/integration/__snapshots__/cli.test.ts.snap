// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli ebs, s3 & ec2 1`] = `
"| Date (UTC)   | EBS Wattage       | EBS CO2e Emissions     | S3 Wattage        | S3 CO2e Emissions      | EC2 Wattage       | EC2 CO2e Emissions     | Sum Wattage       | Sum CO2e Emissions     
| 2020-06-27   | 0.86 Watts        | 0.000611 Kg CO2e       | 1.25 Watts        | 0.000882 Kg CO2e       | 19.29 Watts       | 0.013640 Kg CO2e       | 21.40 Watts       | 0.015133 Kg CO2e       
| 2020-06-28   | 1.73 Watts        | 0.001222 Kg CO2e       | 0.00 Watts        | 0.000000 Kg CO2e       | 0.00 Watts        | 0.000000 Kg CO2e       | 1.73 Watts        | 0.001222 Kg CO2e       
| Total        | 2.59 Watts        | 0.001833 Kg CO2e       | 1.25 Watts        | 0.000882 Kg CO2e       | 19.29 Watts       | 0.013640 Kg CO2e       | 23.13 Watts       | 0.016355 Kg CO2e       "
`;
